### PR TITLE
Add HttpUrl.Builder.addPathSegments()

### PR DESCRIPTION
**Pull request by NightlyNexus**

### CODE REVIEW

**Comment by NightlyNexus** [comment]
I'm probably missing things here, especially in regards to `addEncodedPathSegments()`.

File: okhttp/src/main/java/okhttp3/HttpUrl.java [hunk 1]
```diff
@@ -772,6 +772,23 @@ public Builder addPathSegment(String pathSegment) {
...
     }
 
+    public Builder addPathSegments(String groupedPathSegments) {
+      if (groupedPathSegments == null) {
+        throw new IllegalArgumentException("groupedPathSegments == null");
+      }
+      int startIndex = 0;
+      int endIndex = groupedPathSegments.indexOf('/');
+      while (endIndex != -1) {
+        String pathSegment = groupedPathSegments.substring(startIndex, endIndex);
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/HttpUrl.java]: Oooh, you can avoid the substring call. Push will do that for you automatically.

File: okhttp/src/main/java/okhttp3/HttpUrl.java [hunk 2]
```diff
@@ -772,6 +772,23 @@ public Builder addPathSegment(String pathSegment) {
       return this;
     }
 
+    public Builder addPathSegments(String groupedPathSegments) {
+      if (groupedPathSegments == null) {
+        throw new IllegalArgumentException("groupedPathSegments == null");
+      }
+      int startIndex = 0;
+      int endIndex = groupedPathSegments.indexOf('/');
+      while (endIndex != -1) {
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/HttpUrl.java]: With `Util.delimiterOffset()` it should be straightforward to call `push()` only once within the loop. Something like

```
int offset = 0;
do {
  int segmentEnd = delimiterOffset(...)
  push(groupedPathSegments, ...);
  offset = segmentEnd + 1;
} while (offset <= groupedPathSegments.length());
```
I think one interesting test case will be when the input ends with `/` and I think the resulting URL should then end with `/`. We should probably also confirm that the code works if the input starts with `/`.

Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/HttpUrl.java]: (starting with `/` means there’s going to be an empty segment in the output, like `http://example.com//x`, which is weird but legal)

File: okhttp/src/main/java/okhttp3/HttpUrl.java [hunk 3]
```diff
@@ -772,6 +772,23 @@ public Builder addPathSegment(String pathSegment) {
       return this;
     }
 
+    public Builder addPathSegments(String groupedPathSegments) {
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/HttpUrl.java]: rename `groupedPathSegments` to just `pathSegments` ? Awkward to have the method’s name (path segments) disagree with its parameter

File: okhttp/src/main/java/okhttp3/HttpUrl.java (hunk 2 above)
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/HttpUrl.java]: (The code at the bottom of resolvePath is a good place to start)

File: okhttp/src/main/java/okhttp3/HttpUrl.java (hunk 3 above)
Comment (NightlyNexus) [inline okhttp/src/main/java/okhttp3/HttpUrl.java]: Yup, that was dumb of me haha. fixing.
I am a bit nervous about the similarity in the method names. There would be no cost using the "plural" method, just the possibility of hiding bad input.

File: okhttp/src/main/java/okhttp3/HttpUrl.java (hunk 1 above)
Comment (NightlyNexus) [inline okhttp/src/main/java/okhttp3/HttpUrl.java]: Nice! I should have seen that.

File: okhttp/src/main/java/okhttp3/HttpUrl.java (hunk 2 above)
Comment (NightlyNexus) [inline okhttp/src/main/java/okhttp3/HttpUrl.java]: Ah, yeah, clever.
Cool, going to look into those cases.

File: okhttp-tests/src/test/java/okhttp3/HttpUrlTest.java [hunk 4]
```diff
@@ -801,6 +801,28 @@
     assertEquals(3, HttpUrl.parse("http://host/a/b/c").pathSize());
   }
 
+  @Test public void addPathSegments() throws Exception {
```
Comment (swankjesse) [inline okhttp-tests/src/test/java/okhttp3/HttpUrlTest.java]: Nice tests!

**Comment by NightlyNexus** [comment]
Merged as https://github.com/square/okhttp/commit/de0d48866fe25f2d2643b936c9b04d14841c9536

**Pull request closed without merging**